### ISSUE TITLE
APScheduler 4.0 progress tracking

### ISSUE THREAD

**Issue by agronholm** [agronholm/apscheduler#465]
I'm opening this issue as an easy way to interested parties to track development progress of the next major APScheduler release (v4.0).

## Terminology changes in v4.0

The old term of "Job", as it was, is gone, replaced by the following concepts which are closer to the terminology used by Celery:

- Task definition: a uniquely named callable coupled with configuration like maximum number of instances, misfire grace time etc.
- Schedule: binds a trigger with a task definition
- Job: queued work item for an executor (binds to a task definition, and optionally a schedule)

Notice that the terminology may still change before the final release!

## Planned major changes

v4.0 is a ground-up redesign that aims to fix all the long-standing flaws found in APScheduler over the years.

Checked boxes are changes that have already been implemented.

- [X] Async-first design, with support for asyncio, curio and trio (via [AnyIO](https://github.com/agronholm/anyio))
- [X] Static typing friendly ([PEP 561](https://www.python.org/dev/peps/pep-0561/))
- [X] Support for serializers other than pickle
- [X] Broader time zone support, including zoneinfo time zones ([PEP 615](https://www.python.org/dev/peps/pep-0615/))
- [X] Drop support for Python < 3.7
- [X] [Calendar interval trigger](https://github.com/agronholm/apscheduler/blob/master/apscheduler/triggers/calendarinterval.py)
- [X] Stateful triggers
- [X] `threshold` value for `AndTrigger` (resolves issues with contained `IntervalTrigger` instances)
- [X] The interval trigger should start right away and not after the first interval (#375)
- [ ] Persistent store sharing among multiple schedulers (arguably the most needed feature ever for APScheduler)
- [ ] Context-local job metadata information
- [ ] Easy launching of tasks immediately without needing a schedule

Potential extra features I would like to have:

- [ ] Support for tags in task definitions, schedules and jobs
- [ ] "threshold" value for `OrTrigger` (#427)

You will notice that I have dropped a number of features from master. Some I may never add back to v4.0, even if requested, but do voice your wishes in this issue (and this issue **only** – I will summarily close such requests in new tickets). Others have been removed only temporarily to give me space for the redesign.

## Features on the chopping block

- Twisted scheduler (may be usable through the async scheduler if AnyIO ever gets Twisted support)
- Tornado scheduler (just use the async scheduler)
- Gevent scheduler (does not play well with the new architecture)

**Comment by agronholm** [comment]
The master branch is now in a state where both the async and sync schedulers work, albeit with a largely incomplete feature set. Next I will focus on getting the first implementation of shareable data stores, based on [asyncpg](https://pypi.org/project/asyncpg/). I've made some progress on that a while back but got sidetracked by other projects, particularly AnyIO.

**Comment by codingadvocate** [comment]
Regarding Twisted scheduler on the chopping block for APScheduler v4.

My main OSS project is a multi-process app, that spins up many Twisted reactors in those processes, where several of the sub-processes use APScheduler inside the reactor (https://github.com/opencontentplatform/ocp).  What would be a safe replacement scheduler if the twisted version is being removed?

**Comment by agronholm** [comment]
So you run multiple schedulers? Are you sharing job stores among them?

The main reason I'm thinking of dropping (explicit) Twisted support is because it carries a heavy burden of legacy with it. I will play around with it and see if I can make it work at least with the asyncio reactor. If it can be made to work with a small amount of glue, I will take it off the chopping block.

**Comment by codingadvocate** [comment]
Yes, it runs multiple instances of the schedulers - with their own independent job stores.

I understand the need for software redesigns, and I'm certainly not pushing back or trying to make more work for you.  Just trying to understand what the recommendation would be.  Maybe I could fall back to using APS' BackgroundScheduler since I don't spin it up until after the reactors are running?  Either way, I saw the note and want to ensure I follow whatever happens on that one.  

Either way, thank you for the solid project.

**Comment by agronholm** [comment]
Are the jobs you run typically asynchronous (returning Deferreds) or synchronous (run in threads)?

**Comment by codingadvocate** [comment]
The initial setup with creating job definitions is synchronous.  Any updates to previous job definitions or newly created jobs (stored/managed in a DB) occur regularly in an asynchronous manner (LoopingCall that returns a Deferred).  And all the work with job runtime (execution/management/reporting/cleanup) occurs in non-reactor threads.

**Comment by agronholm** [comment]
Ok, so it sounds like the actual job target functions are synchronous, correct? Then you would be able to make do with the synchronous scheduler, yes?

**Comment by codingadvocate** [comment]
If you're saying so, then yes.  I defer to your knowledge there.  I selected with TwistedScheduler since the user guide choosing-the-right-scheduler section said to do so when building a Twisted application. 

I apologize for compounding the response with a question, but it's related.  How is the thread pool and thread count handled if I use something other than the TwistedScheduler?  Will the job run inside Twisted's thread pool, or inside BackgroundScheduler's thread pool?  Do I need to extend both?

Does constructing the BackgroundScheduler with an explicit max_workers count (example below), do anything when it's running inside the Twisted's reactor?

self.scheduler = BackgroundScheduler({
	'apscheduler.executors.default': {
		'class': 'apscheduler.executors.pool:ThreadPoolExecutor',
		'max_workers': '25'
	}
})

**Comment by agronholm** [comment]
> Will the job run inside Twisted's thread pool, or inside BackgroundScheduler's thread pool? Do I need to extend both?

The sync scheduler (including 3.x's `BackgroundScheduler` knows nothing about Twisted's thread pool. The Twisted scheduler in 3.x differs from `BackgroundScheduler` only in that its default executor uses the Twisted reactor's internal thread pool. It doesn't even have async support!

I want to provide first class async support in APScheduler 4.x. If I can do that with Twisted without having to create an entire ecosystem of Twisted specific components, then I'm open to doing that.

**Comment by agronholm** [comment]
I just added a few items to description:

- External workers
- Schedule-level jitter support
- Ability to cancel jobs
- Timeouts for jobs
- Redis as data store
- Zookeeper as data store
- "executor" being renamed to "worker"

**Comment by thedrow** [comment]
What do you think about adding optional [OpenTelemetry](https://opentelemetry.io/) support?

**Comment by agronholm** [comment]
I am open to it, but only as soon as their API stabilizes. As it stands, every beta release breaks backward compatibility. I have more important issues to work on. I don't think v4.0 will have OpenTelemetry support but I will consider adding it to a minor update release once they are in GA.

**Comment by agronholm** [comment]
A lot of progress has been made on the core improvements of v4.0. Vast code refactorings have taken place. The data store system is really taking shape now.

I've added "Failure resilience for persistent data stores" to the task list. It's one of the most frequent deployment issues with APScheduler, so I'm making sure that it's adequately addressed in v4.0.

I'm not sure what to do with the event system. I may rip it out entirely until I can figure out exactly how it should work. I know users will want to know when a job completes or a misfire occurs etc., so it will be implemented in some form at least before the first release.

I will post another comment when I've pushed these changes to the repository.

**Comment by agronholm** [comment]
I hit a snag with the synchronous version of the scheduler. I tried to use the AnyIO blocking portal system to run background tasks but I had to conclude that it won't work that way. I have an idea for that though.

**Comment by jykae** [comment]
@agronholm do you have any estimate when 4.0 would be released?

**Comment by agronholm** [comment]
I had hoped at least for an alpha at this point, but the design problems in the sync version killed the momentum I had. I have not done any significant F/OSS development since. I am still committed to getting 4.0 done, but due to pressure at work I don't think I can work on it before Christmas holidays.

**Comment by williamwhw326** [comment]
@agronholm  How will you make the jobstore can be shared among multiple schedulers?

**Comment by williamwhw326** [comment]
@agronholm How will you make the jobstore can be shared among multiple schedulers?

**Comment by agronholm** [comment]
> @agronholm How will you make the jobstore can be shared among multiple schedulers?

By coordination and notifications shared between schedulers. Notifications are optional but recommended, and without notifications the schedulers will periodically check for due schedules. How all this works is specific to each store implementation.

**Comment by agronholm** [comment]
@ahmet2mir APScheduler 4.0 already has the proper synchronization mechanisms in place.

What's still missing is the synchronous API. I've come to a realization that I cannot simply copy the async API and remove the `async` keywords because cancellation isn't going to work with the sync API, and AnyIO's `BlockingPortal` mechanism (as it is currently) is inadequate for cases where you need to start background tasks. I must address this issue first and then come back to finish the basic APScheduler 4.0 API.

**Comment by agronholm** [comment]
While 4.0 is being worked on, I've gone back to the 3.x branch for a bit and fixed a number of bugs and other annoyances.

**Comment by agronholm** [comment]
Tests on async/sync workers (formely: executors) are passing now, but the sync worker tests are strangely slow and I want to get to the bottom of that before moving forward.

**Comment by agronholm** [comment]
Slowness in worker tests resolved: it was a race condition in which the notification about the newly added job was sent before the listener was in place, causing the data store to wait for the 1 second timeout to expire before checking for new jobs again.

I'll move on to completing the synchronous scheduler code now. I'm also very close to releasing AnyIO v2.1.0 which is a critical dependency for APScheduler 4.

**Comment by thedrow** [comment]
I can't wait...

**Comment by agronholm** [comment]
Tests for both sync and async schedulers pass, but the tests run into delays caused by the new schedule/job notifications not working as intended, plus the sync scheduler tests are causing lots of "Task exception was never retrieved" errors outside of the actual tests which I will have to investigate. Once these issues have been ironed out, I'm considering making an alpha release.

**Comment by thedrow** [comment]
That would be very helpful.

**Comment by agronholm** [comment]
After hours of debugging, I finally figured out that I was needlessly creating a new task group in the scheduler's `run()` method and overwriting the outer task group as a scheduler attribute. The odd errors went away after I fixed that.

**Comment by agronholm** [comment]
I've just pushed a big batch of changes that implement data store sharing on PostgreSQL (via asyncpg) and MongoDB (via motor). There are a lot of rough edges but at least the whole test suite passes now (at least locally – CI seems to have some troubles). In the coming days I'll try to polish the code base to the point where I can at least make an alpha release.

Feel free to try it out, but you'll have to look at the test suite for some idea on how to use it since I haven't updated the docs yet. Also, the database schema will change before the final release (tasks accounting is not currently done) so expect to have to throw out your schedules and jobs.

**Comment by thedrow** [comment]
Is master now usable?

**Comment by agronholm** [comment]
Usable in the sense that basic functionality works, but I wouldn't rely on it for anything remotely important.

**Comment by thedrow** [comment]
jumpstarter is not production-ready yet.

**Comment by christopherpickering** [comment]
Maybe I missed it in the thread, but when you publish will it have a diff package name on pipy? Thanks!

**Comment by agronholm** [comment]
It will have the same name, so if you're concerned, pin your apscheduler dependencies to < 4.

**Comment by agronholm** [comment]
On another note, CI runs now work again. It was a bit of a head scratcher but turns out the culprit for the CI runs freezing was freezegun v1.1.0. I didn't see this locally because v1.1.0 was freshly released and I hadn't updated my dependencies lately. Pinning to v1.0.0 solved the problem.

**Comment by m3nu** [comment]
We use APScheduler in a [Qt app](https://github.com/borgbase/vorta/) to schedule backups. Works well and I plan to revamp that integration soon. So would be good if QTimer could stay. The 3.x implementation was only a few lines, given that QTimer is quite high level.

Is it still feasible to use QTimer in a `Scheduler` subclass and would you consider merging a PR for it? Or would APS 4 work just as well in a Qt app?

**Comment by agronholm** [comment]
It won't be a priority but I will definitely consider this use case.

**Comment by huangsam** [comment]
Looking forward to the fix for https://github.com/agronholm/apscheduler/issues/285 landing in v4.0 (or some other release) 👍

**Comment by agronholm** [comment]
It's been a while since the last update. My development time has mostly been spent on improving the AnyIO project, and this work has yielded the much awaited 3.0 release which will also benefit APScheduler.

My next step is to refactor the current postgresql data store into a new SQLAlchemy 1.4 based store which would work not only with postgresql but also with mysql and sqlite. I will also implement task accounting (keeping track of how many running jobs per task there are, and ensuring that the limits are respected).

One big decision ahead is how to support data stores using synchronous APIs. It may not make a lot of sense to have the synchronous scheduler use async behind scenes if the data store is fully synchronous. And supporting only the above database backends would leave a lot of users in the cold.

**Comment by samh194** [comment]
Hi @agronholm , is there any update on this release or on #256 in particular?

**Comment by agronholm** [comment]
> Hi @agronholm , is there any update on this release or on #256 in particular?


Initial work has landed on job store sharing, but it's lacking critical components:

* task accounting (making sure the max number of concurrent instances of a task is respected across workers)
* event delivery

I've started work on task accounting last week. The idea is that schedules and jobs are linked to "tasks". A "task" contains some configuration parameters like maximum concurrency, statefulness etc., and links to a callable. When you create a schedule or a job, you need to pass to the method either:

1. the ID of a previously created task
2. a callable object
3. a textual reference (modulename:variablename) to a callable (does not need to be importable on the scheduler process if a persistent job store is used)

When a worker acquires a job, the task gets its counter incremented, and when the job is released, the counter gets decremented. The data store will ensure that the total concurrency of a task never goes above the limit.

I'm still working to figure out the exact semantics involved here, like what happens when a worker tries to acquire a task that has all its concurrency slots taken.

**Comment by agronholm** [comment]
Also, to simplify the implementation of shared data stores, I have decided not to require them to deliver events beyond the current process. This decision could be changed if there is high demand for it.

**Comment by agronholm** [comment]
Also, I'm now trying out a design where synchronous and asynchronous data stores have separate interfaces, and if you use a synchronous data store with an async scheduler, it will just wrap it with threads. I may have been trying too hard to deduplicate the code between sync/async, but if I make the synchronous scheduler independent of the async scheduler, things might get easier. I'm not sure this will be the final design choice that I'll take but it's something I'm exploring now.

**Comment by agronholm** [comment]
Tonight I managed to (hopefully) nail down the new event dispatch system which also has tests with 100% coverage. The work on sync/async schedulers and workers is also coming along nicely. I will commit all the new work to master in a big push once it's in a coherent state.

**Comment by agronholm** [comment]
The data store tests are passing again after my latest batch of changes, and the event dispatch system is shaping up really well. I even managed to make the PostgreSQL data store relay its own events through the database server as asynchronous notifications, making it quite suitable for use without an external messaging system. I will certainly tweak the event system further to make sure the whole system is as observable as it reasonably can be.

My goal is also to do away with the PostgreSQL job store (currently in master only) entirely in favor of the async SQLAlchemy store. I would add its notification features to the SQLAlchemy store which would then be used if a compatible driver was detected.

On another note, I decided to rely on [attrs](https://www.attrs.org/) over dataclasses. I tried really hard to love dataclasses, but seeing that you cannot have ANY optional fields in superclasses when subclasses have mandatory fields is just a showstopper problem for me (dataclasses cannot force keyword arguments, unlike attrs). 

Once I get the rest of the test suite to pass again I will push the changes to Github.

**Comment by agronholm** [comment]
The entire test suite passes now and I have pushed the latest changes to Github. There is an intermittent failure related to the memory job store in `test_remove_job()` when being wrapped as an async datastore. I'm tracking down this annoying Heisenbug.

The API currently revolves around context managers, but I'm going to try to refactor it to be more convenient when used in environments that are less context manager friendly.

I didn't get task accounting done in this batch, but it will be my next focus now that the event dispatch system is in a better shape.

**Comment by agronholm** [comment]
Just fixed a bunch of bugs and got CI to pass on all supported Python versions and platforms (except for Windows+Py3.10 where psycopg2 won't compile).

**Comment by agronholm** [comment]
I took a little detour of adding MySQL and SQLite to the testing matrix, and I'm glad I did because it revealed a bunch of problems. MySQL's timestamp columns don't support fractional seconds by default, and I had to refactor the SQLAlchemy data store a bit to add a workaround for this particular vendor (i.e. using a MySQL specific timestamp type that enables fractional seconds) . As for SQLite, it trashed my latest attempt to enable the SQLAlchemy store to forego the operation of marking schedules as "acquired" for a scheduler because it did not honor the row level locking. The context manager based schedule acquisition API might have been problematic for some other data store implementations anyway. But these issues have been sorted out now and the test suite passes again for all data stores. As a side note, MySQL is **considerably** slower here than the other back-ends and I would not recommend it to anyone.

With that out of the way, I was able to add the first implementation of task accounting. Data stores will now keep track of the number of running jobs per task, and not give workers jobs that would raise the total number for that task above its maximum concurrent job limit. This code will need a lot more tests and polish before I can say it's finished, but the basic functionality is there.

I've also refactored most of the classes to use attrs and all of them to use Python 3.10 style type annotations. I hope this doesn't inconvenience anyone (the library still remains Python 3.7+ compatible).

There are still a couple of unsolved problems with task accounting:
1. How would the system allow tasks to have callables that aren't addressable as `modulename:varname`? Lambdas and local functions (functions inside functions) fall into this category. Normally when a task definition is added to the data store, this addition is communicated to other data stores but there is no way to automatically replicate the same task -> function mapping on data stores in other processes. The use case is important for local data stores, however, and it has to share the same API with persistent data stores.
2. What should happen to jobs when their tasks are updated with different parameters? Should the existing jobs be removed? If the new specification allows fewer concurrent jobs, should we cancel queued jobs? Starting from the oldest or the newest? This is a harder problem than the previous one.

**Comment by agronholm** [comment]
I've now implemented an event broker system which should allow any persistent data store to be shared safely.
The following implementations (in addition to the minimal local-only implementation) are present in the current code:

- Redis
- MQTT (via paho-mqtt)
- PostgreSQL (via asyncpg)

This should be the last major component that was missing from the v4.0 design. From here on out it's just a matter of implementing the promised features, tinkering with the design and polishing the outcome. The documentation will also be largely rewritten but only after the code has more or less settled down.

I've done some work on persistent data stores, too. Each field now corresponds to one column in SQLAlchemy, or one BSON field in a MongoDB document. I made this change to enable more granular updates and deserialization error tracking. The downside is more frequent schema updates when/if more fields are added.

**Comment by agronholm** [comment]
I've now implemented the ability to create a job directly from the scheduler, without the need for a schedule. That's one more item checked off the list.

The event system got a big load of refinements done too, including context manager support for subscriptions and support for one-shot subscriptions.

One design aspect I'm still undecided on is whether to store job results in the jobs table/collection itself. Currently there is a separate table/collection for that.

**Comment by agronholm** [comment]
I've just pushed an implementation of schedule-level jitter support. Another item checked off the list.

**Comment by agronholm** [comment]
And another item crossed off: context-local job info, worker and scheduler (where available).

**Comment by agronholm** [comment]
Yesterday I pushed code that adds failure resilience to persistent data stores, with fully configurable parameters and exponential backoff by default. The challenge is that different data stores raise different exceptions when they fail. This is particularly problematic with the SQLAlchemy store. I've tested this against both PostgreSQL and MySQL servers, shutting them down while the scheduler is running, and I've confirmed that the scheduler starts running again if I restart the server within the configured timeout period. Larger scale testing and refinement is needed, however, to make this feature "rock solid".

This marks the completion of the last mandatory feature of APScheduler 4. I'll start implementing the bonus features now and fixing up the scheduler/worker event logic and other issues in the code, and then gradually rewrite the documentation. When that's done, I will make the first pre-release of v4.

**Comment by agronholm** [comment]
Yesterday I [heard](https://github.com/python-trio/trio/issues/173) about OS level mechanisms that allow more sophisticated waiting techniques than what APScheduler has used for far. Up until now, APScheduler has simply calculated the time differential between the next known scheduled fire time and then used `time.sleep()` or similar to sleep until then (or until awakened explicitly). This method works well in most circumstances, but not in the following:

- the host computer is put to sleep
- the system clock is winded forward

In either of these cases, the scheduler would be blissfully unaware of the abrupt changes in the system clock and may thus sleep past the next fire time. These OS level mechanisms are something I'm now looking to implement for the 4.0 final release, but don't expect to see them in the early pre-releases. They could also be impossible to write automated tests for, so I don't know what to do about that.

**Comment by dariosm** [comment]
@agronholm I'm particularly interested in sharable job store as I commented in another issue, and you pointed me to this one since it's planned for `v4`.

A couple of days ago, I started tinkering over `v3.8.1` for a basic sharable mondgodb store support, but then I found out the feature was `checked` in the list above, so I started digging into `master` branch.
Specifically looking for changes in job stores and the scheduler, and I'm very pleased to see the refactoring done to enable `acquire/release jobs on a store` and `events pub/sub` to trigger jobs to executors. I wondered how you managed to "update" the wakeup time on a certain scheduler instance if another instance added or modified a job, and noticed the `_wakeup_event` strategy to handle such scenario. Nice and clean design.

I'm really anxious now for a release! Apologies in advance if this one was answered in any other thread but... is there any chance for `v4` (`beta` maybe) to be released any time this month?

**Comment by agronholm** [comment]
I've stated elsewhere that that my goal is to get a prerelease out before the end of the year, so I'm sticking with that. I'm currently focused on upgrading another project of mine (Typeguard) but once that work is finished (at least for a pre-release), I will switch my focus back to APScheduler and determine which of the above "extra" features will make it to the pre-release. Note that before the final release, I will likely change the database schemas without notice or migration paths so you should not run any serious workloads with it yet. There is currently also a problem with the schedulers themselves which makes them continuously wake up due to a broken response logic to schedule update notifications from data stores. I plan to address this ASAP when I return to this project. One major subproject remaining before the pre-release is updating the documentation, and I didn't want to start that until the code base was feature complete.

**Comment by dariosm** [comment]
> I've stated elsewhere that that my goal is to get a prerelease out before the end of the year, so I'm sticking with that. 

Hi @agronholm , I hope you had a wonderful start of the year. 
Just wondering here if there are any plans to release `v4` anytime soon? (I'm specifically interested in shared job stores, as stated before).
Since I'm in a hurry to resolve the latter, I have to evaluate to replace `apscheduler` and all my work around it, or wait for the new release to check how it works for my particular requirements.

Thanks again for the great work!

**Comment by agronholm** [comment]
I expect to make an **alpha** release this month. It will still lack some features, but the basics should be working. The data store system has been completely overhauled so it will most likely have some glitches which need to be ironed out over a lengthy period. I wanted to get a release out last month but alas, I have so many projects that need attention that I simply didn't have the time. Also, the database schema is not yet stable so at least during the alpha period, you would have to start from a clean slate from time to time if you plan to use persistent data stores.

On the upside, I think I have finally resolved the last kink that prevented the schedulers from working properly. I have not pushed that code yet but I expect to do so in next new few days, once I've had a chance to test it a little bit locally. After that, I will start rewriting the documentation and once that's done, I expect to make the first pre-release. I don't want to make any promises, but this month is the target for 4.0.0a1.

**Comment by bratao** [comment]
@agronholm sorry for the bump. I'm looking for a job scheduler with Anyio support and would love to use apscheduler. Do you still have plans of releasing version 4.0.0a1? ❤️

**Comment by agronholm** [comment]
Sorry for not responding earlier. I've been crazy busy with both paid work, other F/OSS projects and my laptop just broke down. Mental exhaustion also plays a significant part in the lull in my activity with this project. I will get back to it, hopefully this Saturday. My plan still remains the same: fix the schedulers, rewrite the docs and then release 4.0.0a1.

**Comment by rogueinkamp** [comment]
Just stopping by to say thank you. I have held off thus far to avoid noise/spam, but I must say I do appreciate the hard work on this. I can’t speak for everyone but I am empathetic towards the challenges of FOSS. I am very patient on this release. I’m very happy to wait as to not result in burnout.

Apscheduler is probably one of the most powerful and enjoyable pieces of software I’ve had the pleasure of using. It is great you’ve provided this to us to use for free. Thank you and take care.

**Comment by bratao** [comment]
@agronholm Thank you, I love your work. 
Your wellbeing needs to come first, please, take care of yourself! Burnout is something very serious.

**Comment by dariosm** [comment]
> Sorry for not responding earlier. I've been crazy busy with both paid work, other F/OSS projects and my laptop just broke down. Mental exhaustion also plays a significant part in the lull in my activity with this project. I will get back to it, hopefully this Saturday. My plan still remains the same: fix the schedulers, rewrite the docs and then release 4.0.0a1.

@agronholm take care of yourself!

**Comment by agronholm** [comment]
Thank you so much for the support! That really made my day :heart: 

I just pushed a bunch of commits that deliver both the scheduler fix I talked about, and the first piece of updated documentation. I have other documentation pieces in progress locally and I will be pushing them as I am able to finish them.

I've also been looking into a mechanism that would allow me to add back support for frameworks like Qt (via PySide6) or gevent. Particularly Qt support is something I'm inclined to keep from the 3.x series, given the number of questions and issues raised about it here. But this work is unlikely to land before 4.0.0a1, given how horribly late that release is already.

I've also managed to fix a bunch of common annoyances on the 3.x series. I'll be making a new release of that soon.

**Comment by thedrow** [comment]
I'd defer those additional features to 4.1. What we have in master will be good enough to start with for new users.
Older users will wait till the appropriate time to migrate anyway.

**Comment by aditya-yadav-27** [comment]
Any ETAs for V4 release? I appreciate your work 👍

**Comment by jenfredwell** [comment]
+1 for canceling jobs.  Thank you for this great module!

**Comment by agronholm** [comment]
> Any ETAs for V4 release? I appreciate your work +1

I've given too many ETAs that have all flown by, so I'm not doing that anymore. For v4.0.0a1, the goal is to get, at the minimum, the documentation done and the scheduler working for the most common use cases.

On that note, some practical testing by Donald J. Welch and myself resulted in some interesting findings. I was able to make sample ASGI middleware that should work with any ASGI framework. As for WSGI, it turns out that the synchronous API is not suitable for use in that environment and must be modified so that you can just start the scheduler without using a context manager (which must be used on the async side due to AnyIO/trio compatibility, as they enforce structured concurrency).

I have some time to work on APScheduler this weekend.

**Comment by agronholm** [comment]
Another week has gone by, and the user guide rewrite is well underway. I now also have some PoC code to make APScheduler integrate with both WSGI and ASGI web apps. Writing the WSGI code revealed some issues with the synchronous API, however. Given that using the asynchronous scheduler with `async with` is a must (due to its reliance on [AnyIO](https://anyio.readthedocs.io/) task groups, which are based on [Trio](https://trio.readthedocs.io/)), I thought that I could mirror that in the synchronous API too, but it turns out that supporting WSGI that way is impossible. The scheduler has to be started as an import side effect and has to remain running when the module has finished running, making it impossible to mandate its use through a context manager (`with Scheduler() as scheduler: ...`). As soon as I discovered this problem, I changed the synchronous scheduler API so that it can be run in a background thread, with an `atexit` handler that will gracefully stop it once the process is about to exit.

In the process of writing the user guide, I ran into another API design issue. The 3.x design where schedules can be "tentatively" added to the scheduler, and only "really" added to the store when the scheduler is started has been a huge pain point for users, which is why I got rid of that in v4.0. However, I hadn't yet gotten to the point of giving the user an opportunity to update/replace persistent schedules before the scheduler starts running (and potentially processing obsolete schedules). The way it looks like now is that the scheduler will have to start the data store (and the event broker) on demand, if the scheduler is accessed before its main loop is started. This is hopefully the last API change I will have to make before v4.0.0a1, and I hope I won't run into major issues modifying the synchronous data store/event broker design to accommodate this change.

In short, here is an example of how one would use the synchronous scheduler in a WSGI app with a PostgreSQL data store:
```python3
scheduler = Scheduler(SQLAlchemyDataStore.from_url('postgresql+psycopg2://user:pass@host'))
scheduler.add_schedule(my_function, CronTrigger(hour=0, minute='*/30'), conflict_policy=ConflictPolicy.replace)
scheduler.start_in_background()
```
Constructive criticism is welcome, as always.

**Comment by binbjz** [comment]
This is the most elegant scheduler I've ever used, and I've been looking forward to the release of 4.0. All the hard work you've done for all of this.

**Comment by pfernandesviz** [comment]
Hey @agronholm, congratulations and thank you for the awesome work!
Whats left for the v4 release?
Can I use master now "safely" as is?
Cheers

**Comment by p-fernandes** [comment]
Hey @agronholm, congratulations and thank you for the awesome work!
Whats left for the v4 release?
Can I use master now "safely" as is?
Cheers

**Comment by agronholm** [comment]
Mostly just getting the documentation rewrite finished. I have so many projects that I'm working on that there can be a few weeks of inactivity in any given project because I'm focused on something else.

Some enterprising individuals have already started using master, and reported a number of bugs, all of which I've fixed. Nothing has been reported in a while, so if you know what you're doing, and are expecting some rough edges here and there, go ahead!

**Comment by ccope** [comment]
This library looks great, but I'm working in a flask app and would love that start_in_background feature for the synchronous scheduler. Any chance you could put the work-in-progress up in a branch? Also, maybe making an alpha release, even without docs being updated, would be useful for getting more testing+feedback?

**Comment by agronholm** [comment]
I'm again sorry to make everyone wait, but with my summer vacation approaching, it's looking like I will have enough time on my hands to finish the work left for getting the alpha release out before August. The WSGI changes turned out to require much more substantial changes than I had hoped for, but the end result feels like it was worth the effort. The API is cleaner and easier to understand. A downside of this change is that the sync API is no longer completely symmetrical with the async one, but that's something I can live with.

> Also, maybe making an alpha release, even without docs being updated, would be useful for getting more testing+feedback?

Anyone adventurous enough to try it out can just install directly from Github. I won't need to make a release for that. Several people have already done this. As for the WSGI changes, the code is not yet in working order but will be Real Soon (tm). I will push that to master once the tests pass.

**Comment by kai-tub** [comment]
Hey, I am looking forward to those changes!
This is a "volunteer project" and I think there should be no need for you to apologize. :)
I think everybody will appreciate the new API and be thankful for all of the work you've put in :+1:

> Anyone adventurous enough to try it out can just install directly from Github. I won't need to make a release for that. Several people have already done this.

I would like to point out a different reason to have alpha releases available:
It allows users to publish PyPI packages that depend upon the alpha releases, which is not possible with pointing to the GitHub master branch.
At least I am not aware of how this can work (maybe I didn't do enough research?).
This is maybe a niche use case, but I like to work with alpha releases during the development and test phase and use those when distributing alpha releases of my projects.
I just wanted to add this to give you a potential reason why some people might find alpha releases helpful.

With the upcoming alpha release or not, I am grateful for all of the work you have put into this project.
Thank you :pray:

**Comment by cfaaron** [comment]
> I'm again sorry to make everyone wait, but with my summer vacation approaching, it's looking like I will have enough time on my hands to finish the work left for getting the alpha release out before August. The WSGI changes turned out to require much more substantial changes than I had hoped for, but the end result feels like it was worth the effort. The API is cleaner and easier to understand. A downside of this change is that the sync API is no longer completely symmetrical with the async one, but that's something I can live with.

Here here - many of us lurkers appreciate all you put into this project and your patience and diligence replying to questions and people seeking advice.  Thank you for all of your effort that has created a very useful tool for all of us.  Enjoy your vacation.

**Comment by agronholm** [comment]
> It allows users to publish PyPI packages that depend upon the alpha releases, which is not possible with pointing to the GitHub master branch.

Respectfully, I feel this is not a great idea at this point when the API is still in a flux. Patience, please :)

**Comment by asihacker** [comment]
My project is waiting for your release

**Comment by asihacker** [comment]
Can version 4.0 be released before August 2022

**Comment by agronholm** [comment]
Certainly not the final version. 4.0a1, probably.

**Comment by agronholm** [comment]
I just pushed a bunch of commits that enable the WSGI use case, along with the documentation for that feature. Next, I'll be working on updating the documentation for the 4.0a1 release.

**Comment by agronholm** [comment]
Today I pushed a big batch of commits that updated the [examples](https://github.com/agronholm/apscheduler/tree/master/examples) directory. This took much longer than anticipated (4 days of work) because I kept finding (and fixing) new bugs and API inconsistencies. But things are looking much better now, and it looks like I can finally focus on updating the actual documentation. I still have until the end of the week to deliver the 4.0a1 release :)

**Comment by agronholm** [comment]
Good news for experimenters and early adopters: Hynek Schlawack graciously just made a new release of [attrs](https://www.attrs.org/en/stable/) which enabled the APScheduler test suite to pass on Python 3.11.

**Comment by binbjz** [comment]
> Good news for experimenters and early adopters: Hynek Schlawack graciously just made a new release of [attrs](https://www.attrs.org/en/stable/) which enabled the APScheduler test suite to pass on Python 3.11.

This is really good news. Hopefully, when  python 3.11 is released, our 4.0 will be released as well.

**Comment by agronholm** [comment]
> > Good news for experimenters and early adopters: Hynek Schlawack graciously just made a new release of [attrs](https://www.attrs.org/en/stable/) which enabled the APScheduler test suite to pass on Python 3.11.
> 
> This is really good news. Hopefully, when python 3.11 is released, our 4.0 will be released as well.

No promises. But the documentation is coming along nicely, and still on track for the end of the month target for the first alpha. After that, I expect to add in the bonus features like cancellation, tags etc. along with fixing bugs encountered by the users. Once all the features are in and the bug hunt calms down, I'll make a beta release. How long this will take is anybody's guess.

**Comment by agronholm** [comment]
So, it's August already. I didn't quite meet my self imposed deadline, but if you look at the commit history, I haven't exactly rested on my laurels either. The examples were updated, docstrings written, module structure refactored and all parts of the documentation other than the user guide and the changelog are more or less up to date for v4.0.0a1 now. I just need one or two more productive evenings to get this thing to the goal.

**Comment by asihacker** [comment]
come on.

**Comment by agronholm** [comment]
> come on.

?

**Comment by taybin** [comment]
> come on.

@asihacker Since there is absolutely no reason @agronholm couldn't just say "forget this" and walk away, even when feeling frustrated, a minimal amount of politeness is necessary.  Especially considering that he *is* actively working on and making good progress on 4.0.

**Comment by Jackeriss** [comment]
> 

well, it seems there is some misunderstanding here, "come on" maybe he means “fighting!”, which is "加油" in Chinese.

**Comment by Jackeriss** [comment]
Maybe "fighting" is still not the right translation, it's a little hard to translate "加油"（Add oil）, but 
 I think he just wanted to encourage you. 

https://www.wikiwand.com/en/Add_oil
https://translate.google.com/?hl=zh-CN&sl=zh-CN&tl=en&text=%E5%8A%A0%E6%B2%B9%0A&op=translate

**Comment by taybin** [comment]
@asihacker @Jackeriss In that case, I apologize for the message above.  "Add oil" is a pretty cool idiom.

**Comment by asihacker** [comment]
> 
It means encouraging you

**Comment by asihacker** [comment]
I have been following and learning your project

**Comment by asihacker** [comment]
> come on.

“come on”It means encouraging you

**Comment by asihacker** [comment]
> > come on.
> 
> ?
I have been following and learning your project

“come on”It means encouraging you

**Comment by agronholm** [comment]
I've overcome my writer's block on the user guide and the work is proceeding quite well. Several sections have already been completed, and I'm hopeful that I can get more work done on it tomorrow.

On another note, another prospective user was asking me for assistance experimenting on v4.0 and in the course of lending that assistance, I remembered that right now, job results are always stored in the data store so a schedule would quickly fill the data store with results that nobody cares about. To fix this, I will have to add a new job parameter that specifies an expiration time for the results. A schedule should create jobs whose results are never saved, but that raises another question: how to deal with exceptions occurring when the jobs are run? Should those be implicitly logged by the worker? What about job completion events? Should those contain the exception object too (I'm concerned about what happens if serialization fails)? I'd be grateful for your opinions and insights.

**Comment by rogueinkamp** [comment]
Personally I always want a job exception to be logged by the worker by default. That's leaving a nice "paper trail" for me to start hunting down why my jobs failed. 

I think job completion events sure can contain the exception as well. I'm not fully sure of the implications of this, but it may be nice to have this available. I've logged job exceptions to the database or done other things with them upon job completion handler (like start a different job), so it's nice to have an exception object there. If serialization fails, once again, a nice log message stating that serialization was not successful while also logging the job exception seems sufficient. I like clues when things go wrong. 😀 

I'm not entire sure your I understand your question or if it was open ended to whomever but those are my two cents on the matter 😄

**Comment by agronholm** [comment]
Thanks, that was exactly the kind of feedback I was looking for. I don't 100% agree on everything, so here's what I'm going to do in the short term:

- Add an expiration time field to jobs
- Scheduled jobs have a **default** expiration time of 0, meaning their results are by default never recorded in the data store
- Manually triggered jobs have a reasonably long expiration time, say 1 hour
- The termination of a job is always logged by workers, regardless of why it happened
- The `JobReleased` event will contain the fully qualified name of the exception class, the string representation of the exception object and the traceback, in separate fields

Housekeeping tasks need to be added to the data store system to clean up expired job results, but I will defer that to the second alpha release.

**Comment by courtland** [comment]
I actually like your changes better, particularly with the memory store. The latest master branch is working well for me so far.

**Comment by agronholm** [comment]
Ok, got a lot more work done on the documentation tonight. The user guide is more or less in shape for the first alpha. There are still issues like plenty of trigger documentation having been wiped out in the process of the v4.0 docs rewrite – something I need to fetch back from the 3.x series. But the way I see it, I only need one more push to sort out the minor stuff and push 4.0.0a1 out the door – hopefully as early as tomorrow.

**Comment by agronholm** [comment]
After a bunch of last minute documentation fixes, including another change in the Sphinx theme (`haiku` didn't render a TOC which I found problematic), I've released v4.0.0a1 at long last. Rejoice! :confetti_ball:

One more change I've made is converting the issue templates to use the newer GitHub forms. I had grown tired of people just wiping away the template text and writing whatever they wanted instead. This change will enforce some structure on the bug reports and feature requests and should hopefully act as a deterrent against such behavior.

**Comment by agronholm** [comment]
I intend to take a short break from APScheduler development, but my plan for 4.0.0a2 is:

- Change the scheduler API to require an explicit call to start schedule processing, in order to allow users a chance to update tasks and schedules to match code changes before the scheduler attempts to process/run them
- Implement internal housekeeping (pruning expired job results)
- Add methods for manual task configuration
- Support lambdas and other unreferenceable callables as task functions
- Add an async version of the Redis event broker
- Implement tags, job cancellation and stateful jobs
- Resolve potential inconsistencies with CronTrigger and weekdays (#608)

Let me know if you encounter showstopper issues in your testing. I will try to prioritize such issues.

Planned action items for further releases:

- Find a way to support at least Qt, possibly other event loops
- Implement job timeouts and retries
- Make event broker implementations resistant to connectivity failures, like data stores are now
- Resolve DST bug in CronTrigger (#529, #606)

When the project reaches feature completeness, I will transition it from "alpha" to "beta". At that point, the API churn should significantly slow down.

**Comment by agronholm** [comment]
I've released v4.0.0a2 mainly due to a serialization issue I found where the `JobReleased` event did not serialize properly using JSON or CBOR.

That wasn't the only change though – I did a lot of work on event brokers, making them resistant to connection failures. Another change to take note of is that using a scheduler as a context manager does **not** start it anymore. You need to start it explicitly by calling either `run_until_stopped()` or `start_in_background()`. This change was made to give users an opportunity to manipulate schedules and tasks before the scheduler attempts to process any schedules. An async version of the Redis event broker was also added.

**Comment by agronholm** [comment]
I've been toying with the idea of ditching the synchronous implementations of scheduler and worker and making them wrappers for their asynchronous counterparts. The wrappers would run asynchronous event loops behind the scenes. This would let me eliminate a lot of duplicate async/sync code. The wrappers would still work the same way for the users.

Can anyone think of a reason why I shouldn't go ahead with this?

**Comment by mbrav** [comment]
@agronholm Besides of the possible overhead of running an asynchronous event loop for when it is not necessary, cannot think of anything else. Seems like it would be worth the simplification for code maintenance and removal of duplicate code, but maybe not, if one is purist.

**Comment by kai-tub** [comment]
@agronholm for wrapping async code, I had a very good experience with [asyncer](https://asyncer.tiangolo.com/) from Sebastián Ramírez (the author from FastAPI).
I am not familiar with the source code, so I apologize if it isn't a good fit :)

**Comment by agronholm** [comment]
Thanks for the feedback!

> @agronholm for wrapping async code, I had a very good experience with [asyncer](https://asyncer.tiangolo.com/) from Sebastián Ramírez (the author from FastAPI).

Thanks for letting me know that this exists. I'm not sure it will be useful in this particular project but it does look intriguing.

**Comment by jamesdow21** [comment]
> I've been toying with the idea of ditching the synchronous implementations of scheduler and worker and making them wrappers for their asynchronous counterparts. The wrappers would run asynchronous event loops behind the scenes. This would let me eliminate a lot of duplicate async/sync code. The wrappers would still work the same way for the users.
> 
> Can anyone think of a reason why I shouldn't go ahead with this?

Potential issue could be how it was handled if someone is using APScheduler from within synchronous code that is itself called by asynchronous code  

I've run into an issue where I have written some asynchronous data retrieval (using aiohttp) embedded within a module that was just calling `asynio.run()`   

That worked great until I was trying to use the module within a Jupyter notebook, whihc is running it's own async event loop through Tornado  

Somewhat ugly workaround was essentially to wrap the `asyncio.run()` call in a try..except and if it hit the RuntimeError from an already running event loop, start a new thread and call `asyncio.run(...)` on my coroutine there instead

**Comment by agronholm** [comment]
> Potential issue could be how it was handled if someone is using APScheduler from within synchronous code that is itself called by asynchronous code, since only one event loop can be running at a time

I'm trying to wrap my head around this. With the synchronous wrapper I'm building, the scheduler (and internal worker) would run in a separate thread which then hosts an event loop. If you see a problem occurring in this scenario, could you elaborate on it?

**Comment by econeale** [comment]
Hey @agronholm! Don't want to get off topic here but we're really excited about the store sharing among multiple schedulers - anything low hanging fruit that we could help with to get 4.0 over the finish line?

**Comment by agronholm** [comment]
> line?

I'm not sure about "low hanging fruits" but there **are** things you could do to expedite the process:

1. Test the various external services (PostgreSQL, MySQL/MariaDB, Redis and EMQX) against the corresponding data store and event broker implementations and see how they fare in various situations where connectivity is temporarily lost (timeout, crash, network disconnection etc). These issues should not cause the scheduler to crash (unless they persist for a long time).
2. Draft technical specifications for the "extra features" at the top of this issue. Describe how you would expect them to work. The more detailed the description, the better (within reason of course).
3. Test the latest code in a development environment and report any issues you find

Number 1 and 3 in particular would help build confidence in the system.

**Comment by jamesdow21** [comment]
> > Potential issue could be how it was handled if someone is using APScheduler from within synchronous code that is itself called by asynchronous code, since only one event loop can be running at a time
> 
> I'm trying to wrap my head around this. With the synchronous wrapper I'm building, the scheduler (and internal worker) would run in a separate thread which then hosts an event loop. If you see a problem occurring in this scenario, could you elaborate on it?

As long as it's already in a separate thread, you should be all set.  
I was trying to start a new event loop within the current thread, and then falling back to starting a new thread only if needed.

**Comment by agronholm** [comment]
So, I've been refactoring the code for the next release. There are BIG changes coming:

- The sync scheduler has been turned into a wrapper on top of the async scheduler, running in its own event loop thread (as discussed above)
- The scheduler is now the **only** user-facing interface, meaning that workers can't be started separately without a scheduler, **but** the scheduler can be configured not to process schedules (for whatever reason). This allows me to nix the synchronous worker entirely.
- The synchronous interfaces for data stores and event brokers have been eliminated and the existing synchronous implementations have been migrated to use the async interface with worker threads instead
- The initialization has been changed somewhat. The data store implementation is now responsible for starting up the event broker. A base class has been provided to make this more convenient.

All in all, this has enabled me to eliminate **a lot** of code, further boosting my motivation to work on this project.
If you're not okay with some of these changes, please let me know. Note that I have not yet pushed any of these changes to GitHub.

**Comment by StigKorsnes** [comment]
Any chance we get to see v4 in 2023?

**Comment by agronholm** [comment]
There's always a chance :smiley: 
Right now, the situation is that I'm brewing two major releases in three other projects, and only after that can I get back to APScheduler.

**Comment by rserv-admin** [comment]
Just want to check when we will see the APS v4 coming out

**Comment by agronholm** [comment]
I'm currently on my summer vacation, meaning more time to dedicate to projects. I have a few releases to make in other projects, but APScheduler 4.0 is next on the queue after that. There are great things in the pipeline, including restoration of QT support (I already have a working example locally).

**Comment by agronholm** [comment]
I will make a beta release once the code changes have settled down. Before that, I feel that at least one alpha release is needed.

**Comment by agronholm** [comment]
My summer vacation has come to an end, but I managed to at least get those other releases out of the way. I've pushed some commits to `master` to upgrade dependencies and to shift the range of supported Python versions to 3.8 – 3.12. I'll start working on the next APScheduler release in the coming week.

**Comment by simon-sk** [comment]
Hi @agronholm,
Thanks a lot for your continued work on APScheduler 4.X! 🙌

For a project we are currently working on we would need the “Persistent store sharing among multiple schedulers” functionality of APScheduler 4.X. We found a way to implement it with APScheduler 3.X through database locking but our solution is not as sophisticated as your implementation in version 4.X and we would like to avoid a later migration to 4.X if possible.

Therefore, I wanted to ask if you have a rough estimate how stable the current master branch is and if we can expect a lot of breaking changes with future 4.X releases? We are fine with some rough edges and would also perform extensive testing for our use cases but would like to avoid bigger migration efforts in the future.

**Comment by agronholm** [comment]
APScheduler 4.0 is currently in the alpha phase, so breaking changes will be expected. Once it enters beta, I'll try not to add breaking changes unless deemed essential.

I will later provide a migration path from 3.x to 4.0 by exporting from 3.x and importing to 4.0. It should at least work for built-in triggers.

Thus, sticking with 3.x and later migrating by export+import would seem like the safest option for you.

**Comment by itomatik** [comment]
Thanks for your continued support for APScheduler, @agronholm! What's your rough time horizon on when you'd like to get beta version out? Is it within next couple of months or is it more like next year?

**Comment by agronholm** [comment]
> If we are not using any of the persistence features, but instead solely using the in-memory scheduling, does this warning apply to us? Since we are not storing the scheduled tasks anywhere, I am assuming we won't have to worry about migrating data, so any migration would just need to be for the API (which I think should be easy enough).
> 
> Am I wrong in my assumptions?

Your assumption is correct; if you pin the version, you can choose when you want to deal with the API breakage from an updated version.

**Comment by agronholm** [comment]
> Thanks for your continued support for APScheduler, @agronholm! What's your rough time horizon on when you'd like to get beta version out? Is it within next couple of months or is it more like next year?

Definitely within the next couple months. But given that 4.0 isn't still up to production quality, I decided to put in a little bit of effort to make the 3.x branch a bit less of an annoyance to work with. I even accepted a feature PR (an exceptionally well prepared one) there.

The extent of the changes I've made in `master` since v4.0.0a2 probably necessitate another alpha release, but unless there are major issues with it, I can move things along to beta in short order afterwards. I'm also working on export/import functionality where you can export your jobs in v3 and import them in v4. That should make migration much less of a hassle.

**Comment by agronholm** [comment]
I've pushed the Qt executor to `master` now, thus resurrecting Qt support.
Before I make a new release, I have a few tasks left:

- Add tests for the new executors
- Strip out any traces of features that are not going to make v4.0.0 final
- Go over the code and documentation to make sure it's consistent with the changes since v4.0.0a2

I plan to add the aforementioned import/export features during the beta cycle.

**Comment by binbjz** [comment]
Expect to see 4.0 this year, it's been years.

**Comment by agronholm** [comment]
I just pushed a huge number of documentation fixes, plus some refactorings. The `tags` field had to go for now, but I intend to add it back later. More documentation updates are still needed, but at this rate I expect to make a new release this week or the next. Whether that will be a beta, or another alpha, depends on how I judge the readiness level of the code.

**Comment by agronholm** [comment]
I've just released v4.0.0a3. I took a look at the code base and saw that there are at least three issues that kept me from calling this a beta release:

- There is no way to explicitly configure tasks yet
- There is no clean-up of obsolete job results
- I'm not sure if scheduling class methods is properly working yet

On another note, I would like to invite anyone reading this to fill out [this survey](https://docs.google.com/forms/d/1he3fvGwcwGS9YtRyuFdIAv1kUFuJbGmJNiHAknYwXHA/edit) which will affect the focus of the development of this project.

**Comment by rafalkrupinski** [comment]
> On another note, I would like to invite anyone reading this to fill out [this survey](https://docs.google.com/forms/d/1he3fvGwcwGS9YtRyuFdIAv1kUFuJbGmJNiHAknYwXHA/edit) which will affect the focus of the development of this project.

Regarding question two (thread pool vs event loop thread): 
I'm not sure if I fully understand implications of either answer.
Ideally it would be configurable, like any other element in 3.x branch - schedulers, store etc.

the cases are
- running a small (ending quickly) jobs
    - in thread pool - unnecessary overhead related to threads and serialization (but job store already introduces serialization)
    - in event loop - no issue
- running long jobs
    - in a thread pool - no issues
    - event loop thread - blocking the thread, and I don't think there's any way around it, since jobs are serializable. Unless there's some way to pass an process-local context, which would be useful to for things like thread or connection pools.

So, I've already answered event loop thread, but if the above is right, thread loop seems more logical option, if we have to choose.

**Comment by agronholm** [comment]
It's certainly configurable, but the question is, what should happen *without* any explicit configuration.

**Comment by rafalkrupinski** [comment]
In this case I'd stay with my first answer - event thread loop. Explicit is better than implicit.

**Comment by agronholm** [comment]
2/3 of the survey answers seem to agree with you. I will change the default in the next release.

**Comment by agronholm** [comment]
> 2/3 of the survey answers seem to agree with you. I will change the default in the next release.

I think I spoke to soon. Now with 35 responses, 56.3 % feel that synchronous callbacks should be run in the event loop thread, so I'm holding back that change for the time being.

**Comment by agronholm** [comment]
> running a small (ending quickly) jobs
> * in thread pool - unnecessary overhead related to threads and serialization (but job store already introduces serialization)

I wanted to address this: Running a job in a thread pool would not require serialization. Running in a process pool would.

**Comment by NixBiks** [comment]
It is possible to use the current state of v4 or is that not recommended?

**Comment by agronholm** [comment]
The biggest problem with the current alpha release is that completed jobs are not cleaned up, so the data store gradually fills up with job results. I will address this in time for the next release (4.0.0b1) which I hope get out within a few weeks.

**Comment by camsmith** [comment]
Thanks for working on this @agronholm ! It's a really helpful project. I'm picking it up right now so I was using both v3 and v4 a little bit to test the readiness.

I think the explicit Task creation is preventing use of v4 for me as right now it's unclear to me how to use textual references to functions without manually making a Task. And the task constructor doesn't have defaults so, it's just a bit tricky as is

**Comment by NixBiks** [comment]
> I'm aware of this issue. FastAPI's inability to upgrade its dependencies is due to a [backwards incompatible change](https://github.com/encode/starlette/pull/2026) made in Starlette 0.28. I've been trying to engage with both projects to get this sorted out, but my attempts at getting any useful responses has been in vain so far. There is some ongoing work to this end in [this PR](https://github.com/tiangolo/fastapi/pull/9636).

Ah yeah I noticed you discussed with tiangolo about it [here](https://github.com/tiangolo/fastapi/discussions/10187). My workaround for me, as of now, is to pin fastapi to 0.103.0

**Comment by agronholm** [comment]
> I think the explicit Task creation is preventing use of v4 for me as right now it's unclear to me how to use textual references to functions without manually making a Task. And the task constructor doesn't have defaults so, it's just a bit tricky as is

I made a huge amount of progress to this end last weekend, by implementing the `configure_task()` and `get_tasks()` scheduler methods. I've already updated the documentation locally too, and the only thing left is to update the tests to match.

**Comment by NixBiks** [comment]
v4 is looking really good. I like the data model for the data store!

What is the tags feature exactly? I'm wondering if it's related to a feature that I'm looking for - allow workers to only run certain jobs, similar to [Celery Queues](https://docs.celeryq.dev/en/stable/userguide/workers.html#queues)

**Comment by NixBiks** [comment]
Great @agronholm - you don't have an issue working on it, right?

I might be able to contribute

**Comment by agronholm** [comment]
> Great @agronholm - you don't have an issue working on it, right?
> 
> I might be able to contribute

Help would be appreciated in the planning phase. Writing down just how tag matching should work would be great. For example, do we allow operators like < or > for the fields? Let's say we want to tag a job to only run on Python >= 3.11. Would that be a sensible use case, and how exactly would that work?

**Comment by agronholm** [comment]
Speaking of issues, I just created one where we can discuss this further without pinging everybody: https://github.com/agronholm/apscheduler/issues/798

**Comment by NixBiks** [comment]
> Help would be appreciated in the planning phase. Writing down just how tag matching should work would be great. For example, do we allow operators like < or > for the fields? Let's say we want to tag a job to only run on Python >= 3.11. Would that be a sensible use case, and how exactly would that work?

Wouldn't it be simpler and sufficient to do it similar as with redis queues. The moment you create a job or a schedule, you can also add a list of tags for it. And when you start a worker you can provide a list of tags to consume from. We can jump on a call sometime if you want (I'm in Copenhagen so same time zone I think)

**Comment by PSzczepanski1996** [comment]
Hey @agronholm !

I'm very pleased for new release and I'm waiting for the new documentation.
I'm curious because I'm still unaware how to use `add_schedule` / `reschedule_job` methods in new version.
For example I got this call in my old code and I'm still not sure how can I pass arguments in correct way:
```
    scheduler.add_job(
        patch_mysql_connection,
        args=[instance.submit],
        trigger='date',
        run_date=publish_date,
        id=f'newsletter_job_{instance.id}',
        replace_existing=True,
    )
 ```
 
 My question applies also to the job rescheduling e.g.:
 ```
     scheduler.reschedule_job(
        f'newsletter_job_{instance.id}',
        trigger='date',
        run_date=publish_date,
    )
```

How we now pass date argument? Is there a specific trigger that does that? Also what about rescheduling?
Thanks in advance! It's really nice to see new release nowadays.

**Comment by agronholm** [comment]
I'll be happy to help but please create a new Q/A discussion for this. The short answer is that APScheduler 4 no longer uses entry points (this became very problematic for PyInstaller and other standalone packagers). Therefore you need to pass a trigger instance as `trigger`.

**Comment by agronholm** [comment]
So, I've been hard at work on APScheduler again this weekend. In the process, I've implemented task configuration, enabling users to create schedules targeting lambdas and other un-referenceable callables. I also substantially increased test coverage of the scheduler code, which then led me to discover (and fix) some issues. One annoying issue was that MongoDB's datetimes are only accurate to the millisecond, not microsecond, so I had to come up with a workaround to deal with that. I have no illusions about APScheduler's ability to achieve this precision in reality, but the what were my options? I could accept the inaccuracy, but then users would wonder about the discrepancy when their datetimes don't match anymore after coming back from the DB. I probably haven't even thought about all the subtle issues that would create.

Anyway, I'm nearing the point where the latest batch of changes is ready to be pushed, once I've verified that class methods, static methods and instance methods can also be properly used as task callables. This will likely happen on the next weekend. Given the number of breaking changes I've had to make in the process, I'm considering releasing another alpha if there's demand for that.

**Comment by agronholm** [comment]
Just to break the radio silence: I managed to get explicit task configuration done, with support for just about any callable, so that's 2 out of 4 tasks done from [this comment](https://github.com/agronholm/apscheduler/issues/465#issuecomment-1742211811). The reason I haven't pushed those changes yet is that I stumbled onto an annoying [Heisenbug](https://en.wikipedia.org/wiki/Heisenbug) which causes a database connection to sometimes remain in the pool even when the scheduler has been stopped. My debugging efforts so far haven't yielded any answers, but I will continue to try and fix this.

**Comment by agronholm** [comment]
Wow, I finally figured it out! Turns out that my transactional context manager for the SQLAlchemy data store wasn't quite as airtight as I thought: the transaction is started in a worker thread, but the corresponding exit operation is cancelled before it can take place, so the connection is never returned to the pool.

**Comment by rafalkrupinski** [comment]
👍
Why do you need a worker thread for SQLA? 
I mean, I'd imagine it could run in the event loop just like any other async IO operation.

**Comment by agronholm** [comment]
> 👍 Why do you need a worker thread for SQLA? I mean, I'd imagine it could run in the event loop just like any other async IO operation.

You can't do async I/O when you're using a synchronous database driver like pymysql or cx-oracle. I also have reason to believe that async drivers could also be affected by this, although the tests don't show it.

**Comment by NixBiks** [comment]
Question about the state of stateful. I see you've checked stateful triggers but not stateful jobs. What is a stateful trigger? Currently it isn't possible to share state between jobs, right? E.g. a database connection.

**Comment by agronholm** [comment]
> Question about the state of stateful. I see you've checked stateful triggers but not stateful jobs. What is a stateful trigger? Currently it isn't possible to share state between jobs, right? E.g. a database connection.

Stateful triggers contain state which is saved after the trigger is used to calculate new fire times for a schedule. All triggers are stateful in APScheduler 4. This was necessary in order to correctly implement combination triggers (`AndTrigger` and `OrTrigger`).

**Comment by NixBiks** [comment]
> Stateful triggers contain state which is saved after the trigger is used to calculate new fire times for a schedule. All triggers are stateful in APScheduler 4.

Got it. And it isn't possible to share state between jobs on the same worker currently, right? E.g. I want to reuse a database connection for a schedule (and then close it once the schedule is "done"). Maybe it can be done via events now that I think of it.

**Comment by agronholm** [comment]
For schedules, "stateful" means that its jobs retain some internal state which is then saved after the execution of the job. Sharing database connections is out of scope anyway since you can't serialize them.

**Comment by agronholm** [comment]
I've released another alpha, with tons of fixes/workarounds for less capable RDBMS (sqlite, mysql). Explicit task configuration is also in there. As usual, this update requires wiping your data store and starting over.

**Comment by franz101** [comment]
Thank you so much Alex, I just started using 3.x, do you see any specific date around a production ready 4.x release? @agronholm

**Comment by agronholm** [comment]
I'm sure I can get a _beta_ out before the end of the year (I am furloughed most of December so I have plenty of time to work on APScheduler), but production? That depends on what issues come up in testing. Q2/2024? Not impossible at least.

**Comment by franz101** [comment]
Thank you so much for being transparent, really appreciate your community effort <3

**Comment by agronholm** [comment]
Some good news again. I'm making significant progress on the cleanup feature which periodically purges expired job results, and now also finished schedules which are no longer purged right after the last job is submitted to the store. With luck, I can push these changes to GitHub this weekend.

I've also opened two discussions I would like your input on:

1. [Data store extensibility](https://github.com/agronholm/apscheduler/discussions/822)
2. [Serialization and security](https://github.com/agronholm/apscheduler/discussions/817)

**Comment by agronholm** [comment]
The automated cleanup is now in. That's 3 out of 4 blockers completed for the beta release. My idea of a "beta" release is that it's feature complete but may still contain bugs. I would like to get the data store schema settled so that there won't be any need for nuking the data stores after an upgrade to a newer beta. To that end, the first bullet point of my previous comment needs to be addressed ASAP. In the absence of any feedback on that issue, my plan is to introduce dynamic fields and to correspondingly reduce the number of columns to only those that need to be indexed and queried against.

The last blocker is now the implementation of maximum running jobs limits. Ideally there would be two levels of such limits: task and schedule level. The total number of jobs with the same task ID would never be allowed to rise above the task-level limit, and the number of jobs with the same schedule ID would never exceed the schedule-level limit.

The promised import/export feature will likely not land in the first beta, but probably in the second one.

**Comment by agronholm** [comment]
Alright, so we're in 2024 now. I know what I said about the beta, but I got sidetracked by two other projects of mine that needed urgent work on them. There will be another alpha as soon as:

- #803 is resolved
- The `limit_jobreleased_size` branch is merged with tests (prevents serialization error with the asyncpg event broker when the traceback is too long)
- Data stores have code to periodically update their jobs' expiration times so they're not deleted before completion

**Comment by camsmith** [comment]
Hey @agronholm , wanted to ask whether it would be helpful to submit issues for bugs in the 4.0.0 releases at this point, or if it would just be best to wait? I'd like to move my system to 4.0.0 because it has some settings that 3.10.9 does not, but I'm running into some problems here and there. Thanks again.

**Comment by agronholm** [comment]
It might be helpful, but remember that it's still in alpha state for a reason.

**Comment by JBrut22** [comment]
Hey, I am really loving the new version. it is a lot easier to use when compared to the other options available. I am using AsyncScheduler. Diving into the code, I can see why it is a lot of work to get this version released to the world! I will take a look at the issue below and see if I can make some changes. I also noted that the latest commits may address a few of these issues.

I read through much of the discussion, but thought it prudent to add my own thoughts on v4.0.0a4:
- add_job fails to ever run the job, i think this has already been noted. Basically, this should not be used currently...
    - As i can best tell, this is likely because the job does not register a scheduled_fire_time in the database, so it never runs. Not sure what happens if another job gets added from the schedule.
    - If it is a scheduled job you want to move up, you can remove the schedule and then add it again to have it run right now.
    - if you use add_job the above, then it seems to get stuck and never run any jobs at all.
    - I did create a quick workaround, where it will completely wipe the job history using SQL on start up (or using a FastAPI endpoint. This works great if you only want to use schedules and never add jobs.
    - job results only seem to be an option for add_job, so basically it does not work at the moment as far as i can tell. Not really a big issue...

Workaround, only for scheduling; if you are manually running or adding jobs, this will fail to help you.
It deletes all the jobs, sets running jobs to 0 (in case the scheduler failed during the job), and adds 10 minutes to the schedule so the jobs will start again in a bit. works for my purposes, but might not be suitable for everyone. I run this at start up just in case and created an API endpoint to run it in case it hangs, adding scheduler.stop() and scheduler.start_in_background().

of course, i think there is another fix, referenced [here](https://github.com/agronholm/apscheduler/issues/803#issuecomment-1928516652).
'''
def fix_scheduler():
    """
    fix scheduler
    """
    status_code = 200
    try:
        q = """
            DELETE FROM jobs;
            UPDATE tasks SET running_jobs = 0;
            UPDATE schedules SET next_fire_time = next_fire_time + 600 * 1e6;
            """
        with engine.connect() as conn:
            conn.execute(text(q))
            conn.commit()

        msg = 'Scheduler records fixed.'
    except Exception as e:
        logger.error(e)
        status_code = 500
        msg = f'Failed to fix scheduler records: {str(e)}'
    
    return {"message": msg, 'status_code': status_code}

**Comment by agronholm** [comment]
Hi, it's been a while! I just released a new alpha with a **metric ton** of fixes, and a handful of new features too! Importantly, data stores now finally have a clean-up procedure which will remove expired job results and finished schedules. Schedules can now also be paused and unpaused (contributed by @WillDaSilva). This restores a 3.x series feature in an even more powerful form. Kudos to the 3 people who contributed fixes too!

As usual, the data store schemas have changed in a backwards incompatible manner, so you need to start from scratch when updating. This should stop happening once the beta is out.

**Comment by agronholm** [comment]
> Hey, I am really loving the new version. it is a lot easier to use when compared to the other options available. I am using AsyncScheduler. Diving into the code, I can see why it is a lot of work to get this version released to the world! I will take a look at the issue below and see if I can make some changes. I also noted that the latest commits may address a few of these issues.
> 
> I read through much of the discussion, but thought it prudent to add my own thoughts on v4.0.0a4:
> 
> * add_job fails to ever run the job, i think this has already been noted. Basically, this should not be used currently...
>   
>   * As i can best tell, this is likely because the job does not register a scheduled_fire_time in the database, so it never runs. Not sure what happens if another job gets added from the schedule.
>   * If it is a scheduled job you want to move up, you can remove the schedule and then add it again to have it run right now.
>   * if you use add_job the above, then it seems to get stuck and never run any jobs at all.
>   * I did create a quick workaround, where it will completely wipe the job history using SQL on start up (or using a FastAPI endpoint. This works great if you only want to use schedules and never add jobs.
>   * job results only seem to be an option for add_job, so basically it does not work at the moment as far as i can tell. Not really a big issue...
> 
> Workaround, only for scheduling; if you are manually running or adding jobs, this will fail to help you. It deletes all the jobs, sets running jobs to 0 (in case the scheduler failed during the job), and adds 10 minutes to the schedule so the jobs will start again in a bit. works for my purposes, but might not be suitable for everyone. I run this at start up just in case and created an API endpoint to run it in case it hangs, adding scheduler.stop() and scheduler.start_in_background().
> 
> of course, i think there is another fix, referenced [here](https://github.com/agronholm/apscheduler/issues/803#issuecomment-1928516652).
> 
> ```
> def fix_scheduler():
>     """
>     fix scheduler
>     """
>     status_code = 200
>     try:
>         q = """
>             DELETE FROM jobs;
>             UPDATE tasks SET running_jobs = 0;
>             UPDATE schedules SET next_fire_time = next_fire_time + 600 * 1e6;
>             """
>         with engine.connect() as conn:
>             conn.execute(text(q))
>             conn.commit()
> 
>         msg = 'Scheduler records fixed.'
>     except Exception as e:
>         logger.error(e)
>         status_code = 500
>         msg = f'Failed to fix scheduler records: {str(e)}'
>     
>     return {"message": msg, 'status_code': status_code}
> ```

There are tests making sure `add_job()` works as intended. If you have evidence to the contrary, please file an issue with a minimal working example.

**Comment by hmilkovi** [comment]
In current state (no yet beta) it would be great to add **restart worker after X processed jobs** (behaviour same as Gunicorn) to battle memory leaks in third party libs and code.

Reasons why I would like it:
- slow memory leaks often are detected in production and this is easy way to fix when investigation leaks
- we have same in Celery (I know use case is not 1:1)

Desired flow:
1. global counter
2. if counter > X pause processing wait for all running jobs to finish
3. restart worker

Please give feedback on idea and if people like it I offer my help to make a PR.

**Comment by agronholm** [comment]
> In current state (no yet beta) it would be great to add **restart worker after X processed jobs** (behaviour same as Gunicorn) to battle memory leaks in third party libs and code.
> 
> Reasons why I would like it:
> 
> * slow memory leaks often are detected in production and this is easy way to fix when investigation leaks
> * we have same in Celery (I know use case is not 1:1)
> 
> Desired flow:
> 
> 1. global counter
> 2. if counter > X pause processing wait for all running jobs to finish
> 3. restart worker
> 
> Please give feedback on idea and if people like it I offer my help to make a PR.

GUnicorn is an entry point (or a launcher, if you like), so it's fine for it to do restarts at will. APScheduler, however, is merely a library, and it is not acceptable for it to be restarting its host process. If you want such a mechanism, it will have to be implemented elsewhere.

**Comment by hmilkovi** [comment]
I agree it's out of the scope, many thanks for fast response I really appreciate you work.

**Comment by adk23333** [comment]
hi, I just noticed this project. I would like to know if the documentation for v4.0 can be queried now?

**Comment by agronholm** [comment]
What do you mean by "queried"?

**Comment by adk23333** [comment]
> “查询”是什么意思？

sorry, is "view the document"

**Comment by agronholm** [comment]
The documentation is [here](https://apscheduler.readthedocs.io/en/master/). You just had to select `master` from the drop-down menu at the bottom left.

**Comment by adk23333** [comment]
> You just had to select `master` from the drop-down menu at the bottom left.

thanks

**Comment by parthiibank** [comment]
@agronholm, any idea when we move to beta or stable release which can be used for production? TIA

**Comment by agronholm** [comment]
I was already going to release the beta at the end of July, but then some critical issues were reported by users against `master`. Currently I'm busy getting AnyIO v4.5.0 out the door, and after that I will focus on APScheduler once again. I'm afraid to give any time tables, but getting the beta out is my next top priority after AnyIO.

**Comment by parthiibank** [comment]
Thanks @agronholm. We hit a roadblock while we trying to use 3.10 with clustering. We got undesired behaviors while we ran 3 instances with a shared data store. We are eagerly waiting for a stable release. We have done PoCs with master and for our simple usecase, the master is working fine.

**Comment by PraveenKumarM21** [comment]
@agronholm . I'm using Apscheduler 4.0.0a5. I have a problem with it as I using Redis as event broker and mongo has datastore my flask application is running in multiple instances which inturn creates multiple schedule instances. This causes duplication of tasks. Any idea on this??. Thank you

**Comment by agronholm** [comment]
Could you please not ask questions in this thread, and create a new discussion for that? Any messages posted here will notify a lot of people. My suggestion while waiting for the beta is to use the master directly, as it has tons of fixes already in it.

**Comment by agronholm** [comment]
It's been a while again. This weekend I decided to give the old 3.x branch an overhaul because it was causing difficulties for users. I got a lot of work done over there, including some that will help with the eventual 4.x migration:

- `ZoneInfo` time zones now work with the built-in triggers (they are transparently converted)
- APScheduler 3.x is now compatible with later releases of `tzlocal`
- `CalendarIntervalTrigger` was backported to 3.x
- Export/import of jobs is now implemented in 3.x (4.x implementation TBD)

Check out the full details in the [version history](https://apscheduler.readthedocs.io/en/3.x/versionhistory.html).

I realize that a year ago I estimated the beta to be released in December (2023), and that hasn't happened yet. I was going to do it in July but then a flurry of bug reports came in which I deemed necessary to be fixed before the beta, and the mental exhaustion, plus work from a gazillion other projects hit me hard. But I'm still working on it when I can. I have a hard deadline on finishing another major release before the end of the year, but I'll try to dedicate so time towards fixing the APScheduler 4.x issues too.

**Comment by fcarne** [comment]
@agronholm, hope you're doing well! Are there any updates on APScheduler 4.x?

**Comment by lryan599** [comment]
> > 👍 Why do you need a worker thread for SQLA? I mean, I'd imagine it could run in the event loop just like any other async IO operation.
> 
> You can't do async I/O when you're using a synchronous database driver like pymysql or cx-oracle. I also have reason to believe that async drivers could also be affected by this, although the tests don't show it.

Hi @agronholm. 

Is there a worker thread for synchronous database driver in branch3.x?

**Comment by agronholm** [comment]
> > > 👍 Why do you need a worker thread for SQLA? I mean, I'd imagine it could run in the event loop just like any other async IO operation.
> > 
> > 
> > You can't do async I/O when you're using a synchronous database driver like pymysql or cx-oracle. I also have reason to believe that async drivers could also be affected by this, although the tests don't show it.
> 
> Hi [@agronholm](https://github.com/agronholm).
> 
> Is there a worker thread for synchronous database driver in branch3.x?

No – async support was pretty bad altogether in v3.x.

**Comment by agronholm** [comment]
OK, I got a number of last minute fixes in, mostly related to the handling of acquired schedules and jobs in situations where a scheduler with a fixed ID had previously crashed. I'll have to take another look at `configure_task()` which stumped me a while back and caused me to stop working on v4.x for months. If I can figure it out now, I'll fix it but if not, I'll just go ahead and release a new alpha within the next few days.

There's no good story on task versioning and migration yet which is something I want to solve for v4.0 final, or even export/import which already needs to be in the beta. I also want to add tags and a job filtering system for data stores - the combination of which is essential for heterogeneous worker farm deployments.

**Comment by zhu0629** [comment]
well done, i have already migrated to 4.0beta,
Cloud you please release a new beta version?

**Comment by agronholm** [comment]
> well done, i have already migrated to 4.0beta, Cloud you please release a new beta version?

Where did you get 4.0 beta? The series is still in alpha and I haven't released any beta versions yet.

**Comment by agronholm** [comment]
I've just release v4.0.0a6. I'm sincerely hoping I won't have to release any more alphas, and the next release can be v4.0.0b1.

**Comment by a76yyyy** [comment]
Just wondering if apscheduler 4.0 is production-ready by any chance? Or maybe in beta yet?

**Comment by agronholm** [comment]
> Just wondering if apscheduler 4.0 is production-ready by any chance? Or maybe in beta yet? [@agronholm](https://github.com/agronholm)

My previous message here announced v4.0.0a6 (a = alpha). You don't see any newer versions on the PyPI either. Does that answer your question?

**Comment by a76yyyy** [comment]
> > Just wondering if apscheduler 4.0 is production-ready by any chance? Or maybe in beta yet? [@agronholm](https://github.com/agronholm)
> 
> My previous message here announced v4.0.0a6 (a = alpha). You don't see any newer versions on the PyPI either. Does that answer your question?

Sorry if my question caused any misunderstanding! I'm already using the latest alpha version in testing. What I really wanted to ask is: Do you have plans to release a beta or production version in the near future？

**Comment by acffordyce973** [comment]
Is there any further development of version 4? Will there be a beta?

**Comment by agronholm** [comment]
> Is there any further development of version 4? Will there be a beta?

To answer the question directly, the only notable change lately has been the migration of the MongoDB data store to use the asyncio interface of the `pymongo` library. Other than that, there hasn't been substantial progress in a while, but I do intend to get the 4.0 series to the finish line.

I haven't personally had any use cases for APScheduler in many years, hence the...sporadic progress. I don't remember very well what the biggest issues even were with the alpha, other than the lack of a migration tool. I've started work on the v3->v4 migration tool which is a definite requirement for a beta release. Converting the triggers from v3 to v4 is a steep challenge, however, as they are stateless in v3 and v4. If you want to help, then I would welcome real-world exports of v3 jobs to test the tool against, as I have no such data on my own. Just be mindful of any data privacy issues!

Additionally, I currently have no active work contracts, and finding a proper source of income is my biggest priority right now. To that end, I've started expanding my skill set to include Rust, by rewriting the existing C extension in my other project, [cbor2](https://github.com/agronholm/cbor2), to a Rust extension. Meanwhile, I've started Tidelift maintainership with several of my other projects, but this particular project showed [no potential income](https://tidelift.com/lifter/search/pypi/apscheduler) despite having the most GitHub stars among all of my projects. So if you are using APScheduler for commercial purposes, it would really help a lot if you could sponsor my work, in exhange for consultation.

For anyone else, please check the issue tracker and make sure to report any alpha related issues you've encountered if you don't find the problem reported already. I will always let this thread know when there is something new and noteworthy.
